tests/src/erc20.rs

```diff
@@ -1,152 +1,126 @@
+            Ok(maybe_value) => {
+                let value = maybe_value
+                    .into_t()
+                    .unwrap_or_else(|_| panic!("{} is not expected type.", name));
+                Some(value)
+            }
+        }
     }
 
-    pub fn transfer_from(
-        &mut self,
-        owner: PublicKey,
-        reciever: PublicKey,
-        amount: u64,
-        sender: Sender,
-    ) {
-        self.call_indirect(
-            sender,
-            (
-                (method::TRANSFER_FROM, self.token_hash),
-                owner,
-                reciever,
-                U512::from(amount),
-            ),
-        )
+    fn call(&mut self, sender: Sender, method: &str, args: RuntimeArgs) {
+        let Sender(address) = sender;
+        let code = Code::Hash(self.contract_hash(), method.to_string());
+        let session = SessionBuilder::new(code, args)
+            .with_address(address)
+            .with_authorization_keys(&[address])
+            .build();
+        self.context.run(session);
     }
 
-    pub fn balance_of(&self, account: PublicKey) -> u64 {
-        let balance: Option<U512> = self.query_contract(account.to_string());
-        balance.unwrap_or_else(U512::zero).as_u64()
+    pub fn name(&self) -> String {
+        self.query_contract("_name").unwrap()
     }
 
-    pub fn allowance(&self, owner: PublicKey, spender: PublicKey) -> u64 {
-        let allowance: Option<U512> = self.query_contract(format!("{}{}", owner, spender));
-        allowance.unwrap_or_else(U512::zero).as_u64()
+    pub fn symbol(&self) -> String {
+        self.query_contract("_symbol").unwrap()
     }
 
-    pub fn total_supply(&self) -> u64 {
-        let balance: Option<U512> = self.query_contract(key::TOTAL_SUPPLY.to_string());
-        balance.unwrap().as_u64()
+    pub fn decimals(&self) -> u8 {
+        self.query_contract("_decimals").unwrap()
     }
 
-    fn call_indirect(&mut self, sender: Sender, args: impl ArgsParser) {
-        let Sender(address) = sender;
-        let code = Code::Hash(self.indirect_hash);
-        let session = SessionBuilder::new(code, args)
-            .with_address(address)
-            .with_authorization_keys(&[address])
-            .build();
-        self.context.run(session);
+    pub fn balance_of(&self, account: AccountHash) -> U256 {
+        let key = format!("_balances_{}", account);
+        self.query_contract(&key).unwrap_or_default()
     }
 
-    fn contract_hash(context: &TestContext, name: &str) -> Hash {
-        let contract_ref: Key = context
-            .query(account::ALI, &[name])
-            .unwrap_or_else(|_| panic!("{} contract not found.", name))
-            .into_t()
-            .unwrap_or_else(|_| panic!("{} is not a type Contract.", name));
-        contract_ref
-            .into_hash()
-            .unwrap_or_else(|| panic!("{} is not a type Hash.", name))
+    pub fn allowance(&self, owner: AccountHash, spender: AccountHash) -> U256 {
```

Comment:
Reuse this function for allowances keys in contracts.
Do the same for balances.

Reply:
That's not done.

Reply:
The one in tests uses `self.query_contract` which I don't understand how we would reuse it?

Reply:
Changed the keys in main